### PR TITLE
George/ocr poc

### PR DESCRIPTION
**Pull request by GeorgeCodes19**
## Ticket

{LINK TO TICKET}

### CODE REVIEW

File: app/package.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
   "dependencies": {
     "@storybook/addon-links": "^6.0.0",
     "@trussworks/react-uswds": "^5.0.0",
+    "@types/node": "^20.11.24",
```
Comment (GeorgeCodes19) [inline app/package.json:24]: Needed so that TS can stop complaining about accessing the node process `process.env['SOME_VAR']`. Accessing the system environments can be useful for debugging, service setup, etc. I initially had a debug flag for the logger relying on an env variable.

File: app/src/service/factory.ts
Comment (GeorgeCodes19) [inline app/src/service/factory.ts:1]: Initially this logic was spread throughout the codebase, but I've decided to have a single factory service that's responsible for generating entities that the code relies on. As the service grows we can have multiple factories, but I think this one factory file is good for now.

File: app/src/service/ocr/document/adpEarningsStatement.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { createDocumentMatcher } from "@/service/factory";
+
+export const adpEarningsStatement = createDocumentMatcher(
+  "ADP Earnings Statement",
+  "adpEarningsStatement",
+  {
+    company: /(H GREG NISSAN DELRAY LLC)/,
```
Comment (GeorgeCodes19) [inline app/src/service/ocr/document/adpEarningsStatement.ts:7]: I realize that these are currently hard-coded. We can work on defining new regex patterns (and ideally functions) used to extract relevant information from the OCR text results.

File: app/src/service/ocr/document/w2.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+export const w2Form = createDocumentMatcher("W2 Form", "w2", {
+  employerIdentificationNumber: /\b(\d{2}-\d{7})/,
+  wagesTipsOthers: /(\d+,\d+\.\d{2}|\d+\.\d{2})\s+/,
+  federalIncomeTaxWithheld: /(\d+,\d+\.\d{2}|\d+\.\d{2})$/,
+  employeeAddress: /Shawn Bailey\s*(.*?)\s*\d{3}-\d{2}-\d{4}/,
+  ssn: /\b\d{3}-\d{2}-\d{4}\b/,
+  bottomLines: /(Form W-2 Statement[\s\S]*?)$/,
+});
+
+export type W2FormPatterns = typeof w2Form.patterns;
```
Comment (GeorgeCodes19) [inline app/src/service/ocr/document/w2.ts:12]: The `createDocumentMatcher` located [here](https://github.com/DSACMS/iv-doc-uploader/pull/11/files#diff-d51131a73161318ac3462120862ba2944e7cdee72c51a66ac122e45bf3346bebR29) is responsible for adding the patterns attribute.

File: app/src/service/ocr/index.ts [hunk 4]
```diff
@@ -1,35 +1,138 @@
...
   await worker.terminate();
 
   return output;
 };
 
+// a function that rotates an image and processes it for each orientation
+const processDocument = async (
+  document: DocumentImage
+): Promise<ProcessedRotatedImagesResult> => {
+  const processImage = async (
```
Comment (GeorgeCodes19) [inline app/src/service/ocr/index.ts:52]: Typically I don't nest closures within functions because it's hard to test, but for a POC/MVP I figured this small anti-pattern wouldn't hurt.

File: app/src/service/ocr/index.ts [hunk 5]
```diff
@@ -1,35 +1,138 @@
...
 export const logger = createLogger("ocr-parser", {
-  enabled: process.env.LOG_LEVEL === "debug",
+  enabled: false,
   mixin: () => {
     return { module: "ocr-parser" };
   },
 });
 
-const scanImage = async (path: string, opts: OcrOptions) => {
+const getTextFromImagePath = async (
```
Comment (GeorgeCodes19) [inline app/src/service/ocr/index.ts]: I've made a note to change the name of this function to `getTextFromDocument` since it actually runs the OCR against the `data` property of the `DocumentImage` since the `data` property holds a reference (in-memory) to the rotated document image. 

Otherwise we'd have to write a tmp rotated file to disk and pass its path. I've found that working with files that are written/modified/deleted during a short-lived process could lead to file lock issues.

**Comment by GeorgeCodes19** [comment]
### Service Factory [(`app/src/service/factory.ts`)](https://github.com/navapbc/iv-doc-uploader/blob/george/ocr-poc/app/src/service/factory.ts)

#### Logger Creation (`createLogger`)
Implements pino for application logging.

#### Document Matcher Creation (`createDocumentMatcher`)
Generates matchers with regex patterns for document parsing, facilitating the addition of new document types.

#### Document Image Creation (`createDocumentImage`)
Prepares images for OCR using sharp to extract metadata. sharp is also used for changing document orientation.

### OCR Service [(`app/src/service/ocr/index.ts`)](https://github.com/navapbc/iv-doc-uploader/blob/george/ocr-poc/app/src/service/ocr/index.ts)

#### OCR Processing
> Executes OCR on images using tesseract.js. Includes:

- `getTextFromImagePath` - Process a single document with a given orientation.
- `process` - Process a single document.
- `processDocument` - Create a copy of the image in different orientations (90, 180, 270) and pass each document to the process function.

### Document Parsing
Applies regex patterns to OCR text output, extracting predefined data fields.

#### Document Matchers (`app/src/service/ocr/document/*.ts`)
Specifies regex patterns for targeted data extraction from OCR text.

### Utilities [(`app/src/utils/document.ts`)](https://github.com/navapbc/iv-doc-uploader/blob/george/ocr-poc/app/src/utils/document.ts)

#### Document Image and Orientation (`DocumentImage`, `rotateDocumentImage`)
Defines image properties and a rotation function to align text for OCR processing.

File: app/src/service/factory.ts
Comment (allthesignals) [reply app/src/service/factory.ts:1]: Oh, I love this! Nit, but can we call the file "factories"? I think that'd make more intuitive sense because there are more than one.

File: app/src/service/ocr/document/w2.ts (hunk 3 above)
Comment (allthesignals) [reply app/src/service/ocr/document/w2.ts:12]: Love this, thanks for streamlining this into something more generic. This is cool...

I suspect as this evolves, though, these regex patterns will grow, I wonder if each key could support multiple patterns.

File: app/src/service/ocr/index.ts [hunk 6]
```diff
@@ -1,35 +1,138 @@
...
+    text: result.data.text,
+    confidence: result.data.confidence,
+  };
+
   await worker.terminate();
 
   return output;
 };
 
+// a function that rotates an image and processes it for each orientation
```
Comment (allthesignals) [inline app/src/service/ocr/index.ts:48]: Why do we need to rotate the image? Is this a tesseract thing? Or an OCR best practice or something?

File: app/src/service/ocr/index.ts (hunk 5 above)
Comment (allthesignals) [reply app/src/service/ocr/index.ts]: Can we update the name? Within the context of this file, it's a little hard to follow with the naming misdirection (This name suggest the file isn't already written somewhere or in memory).

File: .gitignore [hunk 7]
```diff
@@ -1,4 +1,5 @@
 # MacOS files
 .DS_Store
 # Developer-specific IDE settings
-.vscode/settings.json
\ No newline at end of file
+.vscode/settings.json
+app/tests/var/*
```
Comment (GeorgeCodes19) [inline .gitignore:5]: The test suite is capable of writing the rotated document images to disk (the `app/tests/var` directory) for visual comparison. [This logic is currently commented out](https://github.com/DSACMS/iv-doc-uploader/pull/11/files#diff-36b38b83fc11ae6dffca708ec2150e49bc3701baa03b708c85e4380db554e0dbR47), but may be useful when testing/developing locally.

File: app/src/service/ocr/index.ts [hunk 8]
```diff
@@ -1,35 +1,138 @@
...
+const processDocument = async (
```
Comment (allthesignals) [inline app/src/service/ocr/index.ts:49]: So we already have a function called `process`, but this function is `processDocument`, and it rotates the image then calls the `process` function below... can we name this something a little less general? Is this the main entrypoint of the ocr service?

File: app/src/service/ocr/index.ts [hunk 9]
```diff
@@ -1,35 +1,138 @@
...
+  document: DocumentImage
+): Promise<ProcessedImageResult> => {
+  const { text, confidence } = await getTextFromImagePath(document, {
+    debug: true,
+  });
+  const parsersArray = Object.values(parsers);
+  const docs = parseOcrResult(text, parsersArray, logger);
+
+  // let's keep track of the number of matched fields per parser
+  const percentages: Record<string, number> = {};
```
Comment (allthesignals) [inline app/src/service/ocr/index.ts:83]: I like this, but out of curiosity what is the difference between including match percentages as part of the OCR output and using the logger you set up?

File: app/src/service/ocr/parser.ts [hunk 10]
```diff
@@ -12,67 +27,43 @@ export type ParsedData = {
...
-  const employerMatch = outputText.match(employerInfoPattern);
+export type ParsingFunction = (
+  documentText: string,
+  patterns: DocumentMatcher<ParserKeys, ParsingPatterns>[],
+  logger: typeof ocrLogger
+) => ParsingFunctionResult;
 
-  const employeeAddressPattern = /Shawn Bailey\s*(.*?)\s*\d{3}-\d{2}-\d{4}/;
-  const employeeAddressMatch = outputText.match(employeeAddressPattern);
+export const parseOcrResult: ParsingFunction = (
```
Comment (allthesignals) [inline app/src/service/ocr/parser.ts]: Nit but this could just be `parse`

File: app/src/service/ocr/parser.ts [hunk 11]
```diff
@@ -12,67 +27,43 @@ export type ParsedData = {
...
+  logger.info("Parsing OCR Data");
 
-  const ssnPattern = /\b\d{3}-\d{2}-\d{4}\b/;
-  const ssnMatches = outputText.match(ssnPattern);
+  const documentMatches = documentMatchers.reduce(
+    (
+      acc: ParsingFunctionResult,
+      matcher: DocumentMatcher<ParserKeys, Record<string, RegExp>>
+    ) => {
+      const { id, patterns } = matcher;
```
Comment (allthesignals) [inline app/src/service/ocr/parser.ts:52]: I think I like `patterns` as a name for now, I'd be curious if there's a future where we'd want multiple patterns for the same key...

My instinct would be to call this "fields" or something more specific to help differentiate better 🤷

File: app/src/utils/document.ts
Comment (allthesignals) [inline app/src/utils/document.ts:1]: Could we name this to something more specific? Looks like it provides a rotation function so maybe `image-utilities` or something?

File: app/tests/service/ocr.spec.ts [hunk 12]
```diff
@@ -1,50 +1,203 @@
...
+
+    const expected = {
+      testDocument: {
+        company: "H GREG NISSAN DELRAY LLC",
+        employeeName: "ASHLEY STELMAN",
+        employeeAddress: "1306 HELLIWELL ST NW PALM BAY FL 32907",
+      },
+    };
+
+    const result = parseOcrResult(documentText, documentMatchers, logger);
```
Comment (allthesignals) [inline app/tests/service/ocr.spec.ts]: The DI pattern you used is so cool and makes testing a dream.

File: app/tests/service/ocr.spec.ts [hunk 13]
```diff
@@ -1,50 +1,203 @@
...
+          payDate: "04/28/2023",
+          earnings: "3,286.78",
+          netPay: "2,921.82",
+        },
+        w2: {
+          wagesTipsOthers: "3,286.78",
+        },
+      };
+
+      expect(highConfidenceResults[0].documents).toEqual(expectedDocs);
```
Comment (allthesignals) [inline app/tests/service/ocr.spec.ts:191]: Sweet! I can see us using this e2e test as a way to check other document types.

File: app/src/service/ocr/index.ts (hunk 6 above)
Comment (GeorgeCodes19) [reply app/src/service/ocr/index.ts:48]: So we rotate the image because the user could submit an image of a document that's not in the correct orientation. Tesserect supposedly has the ability to auto-rotate an image, but it didn't work each time I attempted it. Maybe there's an underlying error with the Javascript implementation/wrapper.

Let's say that the image is sent to them via text, but somehow it's not the proper orientation once it's downloaded (I've actually had this happen) and the user isn't sure how to rotate the image. I think most users would try to do this on a computer rather than being aware that they can do so on their mobile device.

File: app/src/service/ocr/index.ts (hunk 8 above)
Comment (GeorgeCodes19) [reply app/src/service/ocr/index.ts:49]: I ran out of gas when it came to naming things haha. Do you have any suggestions? Maybe `processImage` which gets called by `processDocument`? Since a single document has 4 different images that should be processed (one for each orientation).

File: app/src/service/ocr/document/w2.ts (hunk 3 above)
Comment (GeorgeCodes19) [reply app/src/service/ocr/document/w2.ts:12]: I didn't understand what you meant at first, but now I do. That would be pretty cool! I think it'll be easy to implement as well. My thought was that for a specific document - that there wouldn't be multiple patterns for one field. If we decide to support "unknown" documentation (maybe outside the scope of this project) then I think having multiple expressions to identify a single piece of information would be handy.

File: app/src/utils/document.ts
Comment (GeorgeCodes19) [reply app/src/utils/document.ts:1]: I meant for this file to be a placeholder for all document related functions. I can forsee us processing PDFs as well. For the time I think `documentImage` might be explicit enough (since both functions require DocumentImage as an argument). That also brings up another coding covention question... hyphens or no hyphens (i.e. blurDetector.ts vs blur-detector.ts)

File: app/src/service/ocr/index.ts (hunk 9 above)
Comment (GeorgeCodes19) [reply app/src/service/ocr/index.ts:83]: I don't see the logger being vital for anything except for telemetry. Basically viewing/debugging the internals of the application to help troubleshoot. What's being logged won't necessarily make its way to the client. The percentage of fields found for a particular document is something that I feel that a well crafted UI/UX could utilize.
